libraries/lib-stretching-sequence/tests/StretchingSequenceTest.cpp

```diff
@@ -0,0 +1,193 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  StretchingSequenceTest.cpp
+
+  Matthieu Hodgkinson
+
+**********************************************************************/
+#include "StretchingSequence.h"
+#include "AudioContainer.h"
+#include "AudioContainerHelper.h"
+#include "FloatVectorClip.h"
+#include "MockAudioSegmentFactory.h"
+#include "MockPlayableSequence.h"
+#include "SampleFormat.h"
+#include "WavFileIO.h"
+
+#include <array>
+#include <catch2/catch.hpp>
+
+using namespace std::literals::string_literals;
+using namespace std::literals::chrono_literals;
+
+namespace
+{
+constexpr auto sampleRate = 3;
+} // namespace
+
+TEST_CASE("StretchingSequence unit tests")
+{
+   SECTION("Get")
+   {
+      constexpr auto backwards = true;
+      constexpr auto numChannels = 1u;
+
+      SECTION("Samples can be queried backwards")
+      {
+         // Mimic the way a client would use the legacy API to read samples
+         // backwards.
+         // Our sequence : 1s silence, 1s clip, 1s silence, 1s clip.
+         const auto clip1 = std::make_shared<FloatVectorClip>(
+            sampleRate, std::vector<float> { 1.f, 2.f, 3.f }, numChannels);
+         const auto clip3 = std::make_shared<FloatVectorClip>(
+            sampleRate, std::vector<float> { 4.f, 5.f, 6.f }, numChannels);
+         clip1->playStartTime = 1.0;
+         clip3->playStartTime = 3.0;
+         const auto mockSequence =
+            std::make_shared<MockPlayableSequence>(sampleRate, numChannels);
+         const auto sut = StretchingSequence::Create(
+            *mockSequence, ClipConstHolders { clip1, clip3 });
+
+         constexpr auto len = 2;
+         const std::array<std::vector<float>, 6> expected { { { 6.f, 5.f },
+                                                              { 4.f, 0.f },
+                                                              { 0.f, 0.f },
+                                                              { 3.f, 2.f },
+                                                              { 1.f, 0.f },
+                                                              { 0.f, 0.f } } };
+         auto start = 12;
+         for (const auto& expected : expected)
+         {
+            AudioContainer output(len, numChannels);
+            constexpr auto backwards = true;
+            sut->GetFloats(
+               AudioContainerHelper::GetData(output).data(), start, len,
+               backwards);
+            REQUIRE(output.channelVectors[0] == expected);
+            start -= 2;
+         }
+      }
+
+      SECTION("reconstructs segment sequence when expected")
+      {
+         auto factory = new MockAudioSegmentFactory;
```

Comment:
This is only test code, but I don't like naked `new`.  You could `make_unique` and then `move` it.

Reply:
I'd have refer to `factory` after the move. It's feasible but more complicated.